src/Metadata/Api/HookMethodsCollection.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+declare(strict_types=1);
+/*
+ * This file is part of PHPUnit.
+ *
+ * (c) Sebastian Bergmann <sebastian@phpunit.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+namespace PHPUnit\Metadata\Api;
+
+use function array_map;
+use function count;
+use function usort;
+use ArrayObject;
+use Countable;
+use IteratorAggregate;
+use Traversable;
+
+/**
+ * @internal This class is not covered by the backward compatibility promise for PHPUnit
+ *
+ * @implements IteratorAggregate<int, string>
+ */
+final class HookMethodsCollection implements Countable, IteratorAggregate
+{
+    /**
+     * @var non-empty-list<HookMethod>
+     */
+    private array $hookMethods;
+
+    public static function defaultBeforeClass(): static
+    {
+        return new self(new HookMethod('setUpBeforeClass', priority: 0), shouldPrepend: true);
+    }
+
+    public static function defaultBefore(): static
+    {
+        return new self(new HookMethod('setUp', priority: 0), shouldPrepend: true);
+    }
+
+    public static function defaultPreCondition(): static
+    {
+        return new self(new HookMethod('assertPreConditions', priority: 0), shouldPrepend: true);
+    }
+
+    public static function defaultPostCondition(): static
+    {
+        return new self(new HookMethod('assertPostConditions', priority: 0));
+    }
+
+    public static function defaultAfter(): static
+    {
+        return new self(new HookMethod('tearDown', priority: 0));
+    }
+
+    public static function defaultAfterClass(): static
+    {
+        return new self(new HookMethod('tearDownAfterClass', priority: 0));
+    }
+
+    private function __construct(HookMethod $default, private bool $shouldPrepend = false)
+    {
+        $this->hookMethods = [$default];
+    }
+
+    public function add(HookMethod $hookMethod): static
+    {
+        if ($this->shouldPrepend) {
+            $this->hookMethods = [$hookMethod, ...$this->hookMethods];
+        } else {
+            $this->hookMethods[] = $hookMethod;
+        }
+
+        return $this;
+    }
```

Comment:
I made this to mimic the different behavior where certain hook were added with `array_unshift`, and other with array concatenation, in order to keep BC when methods are added without priority

Reply:
Thank you. I cannot tell at the moment when I will have the time to review this. I would like to note the following, though: the status quo is, AFAIR, that there is an implicit priority when multiple hook methods are configured using attributes: "before" hook methods from a parent class before "before" hook methods from a child class, "after" hook methods from a parent class after "after" hook methods from a child class, etc. There is also an implicit order within a class based on the line number of the method declaration (in other words: the order in which the Reflection API finds the methods). This implicit ordering must be preserved.

Reply:
This was more or less what I what trying to explain 😅

Reply:
Hi @sebastianbergmann 

I've added the test [`tests/end-to-end/metadata/hook-methods-order.phpt`](https://github.com/sebastianbergmann/phpunit/pull/5890/files#diff-d87d5bdbabdcb5f9b18e7b4ad9cb06638a6e39057d540dd97369ace116d98565) in order to ensure that this "natural order" is still respected